### PR TITLE
Blazor WebAssembly debugging update

### LINKED ISSUE #15983: Attach debugger to Blazor WebAssembly process
React to https://devblogs.microsoft.com/aspnet/asp-net-core-updates-in-net-core-3-1/

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: ec2f61c7-9f51-6735-e328-3c7b495d70c3
* Version Independent ID: 4d06e565-282d-6387-aa90-cb2c2986d35a
* Content: [Debug ASP.NET Core Blazor](https://docs.microsoft.com/en-us/aspnet/core/blazor/debug?view=aspnetcore-3.1)
* Content Source: [aspnetcore/blazor/debug.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/debug.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #15983
Addresses #16270
Addresses #16273

DR: Most of the discussion below is TL;DR on the `no-loc` situation, which you can probably skip. There's one critical bit (https://github.com/aspnet/AspNetCore.Docs/pull/16123#issuecomment-570317743). Debugging doesn't work on Linux due to the setup in the framework: `Unable to find debuggable browser tab` instructions content always calls `GetLaunchEdgeInstructions` (i.e., on Linux), but `GetLaunchEdgeInstructions` throws for Linux at https://github.com/dotnet/aspnetcore/blob/e72223eaf58a3ee6660a922064d2449e47b78253/src/Components/Blazor/Server/src/MonoDebugProxy/BlazorMonoDebugProxyAppBuilderExtensions.cs#L288.

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/blazor/debug?view=aspnetcore-3.0&branch=pr-en-us-16123)

@VILcryo, you can try the updated steps on this PR, or you can wait until we get this reviewed+merged. There's no ETA at the moment tho on when these changes will go live given that the 🎁 *holidays* ⛄️ are on us and things will be mostly shutting down for the period around here.

### CODE REVIEW

**Comment by guardrex** [comment]
@JasonCard It looks like I need to quote file names. When I had *blazor.webassembly.js* unquoted ...

```
no-loc: [Blazor, SignalR, blazor.webassembly.js]
```

It 💥 the build ... the build just chokes and freezes.

If I quote it ...
```
no-loc: [Blazor, SignalR, "blazor.webassembly.js"]
```

... the build is :+1:. I'll go forward this way unless you think it's a problem. When one or more of these PRs merge, I'll check a non-English doc to see if it worked.

**Comment by JasonCard** [comment]
@guardrex cc @Rick-Anderson Thank you for your help to test the no-loc syntax. However, it looks like [debug.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/debug.md) does NOT have an instance of "blazor.webassembly.js" to see if it gets blocked from localization. Can you please help to test that no-loc metadata in double quotes in an article that contains "blazor.webassembly.js" like [hosting-models.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/hosting-models.md)?

**Comment by Rick-Anderson** [comment]
@JasonCard if you know what needs to change can you make the change? @guardrex  is on vacation for a couple weeks and I'm not sure what needs to change.

**Comment by guardrex** [comment]
The updates have already been made ... to all of the Blazor topics.

WRT it not being in this particular  topic: I'm going to use the same string for all topics, it has been applied to them on other PRs.

I'll keep you posted.

**Comment by JasonCard** [comment]
Oh, okay, I just created this PR [https://github.com/aspnet/AspNetCore.Docs/pull/16325](https://github.com/aspnet/AspNetCore.Docs/pull/16325), but it can be closed if not needed.

**Comment by solrevdev** [comment]
So I followed the procedure from the [.NET Core CLI](https://github.com/aspnet/AspNetCore.Docs/pull/16123/files#diff-0c89e7bfb3d7369c12990f538d018d02R73) part of this document and the issue/exception remains the same. 

For info see the issue I previously raised  #16366

**Comment by guardrex** [comment]
@solrevdev When I set up the steps here, it worked. I'll run it again when I get back next week. btw - Make sure that you're on 3.1. I'm still OOF until Monday ... savoring my last few days off until a **_BIG_** workload hits in January. Let me take a look and get back to you next week.

**Comment by guardrex** [comment]
@solrevdev Interesting ... that's thrown because Linux **_isn't supported_** in the framework ...

https://github.com/aspnet/AspNetCore/blob/e72223eaf58a3ee6660a922064d2449e47b78253/src/Components/Blazor/Server/src/MonoDebugProxy/BlazorMonoDebugProxyAppBuilderExtensions.cs#L288

See the conditionals there ... `OSPlatform.Windows` and `OSPlatform.OSX`.

@danroth27 can let us know if an issue should be opened for engineering to provide a Linux instruction there **_IF_** Linux really is supported for debugging. For the time being tho WRT you actually getting it to work, no 🎲🎲 dice on that right now ... your only options there are Windows and macOS.

**Comment by solrevdev** [comment]
Thanks so much for looking into this especially seeing it’s still the holidays!

**Comment by JasonCard** [comment]
Hi @guardrex, Xamarin Docs had similar issue with "Xamarin.Forms", so they tested the no-loc metadata to see if "Xamarin.Forms" can be locked. However, it did not work even when surrounding the terms with double quotes... So currently the only way to use no-loc feature to lock terms connected by periods like "blazor.webassembly.js" may be to use inline syntax like `:::no-loc text="blazor.webassembly.js":::` for every instance. But I don't expect your team to do this as it's not a scalable solution. We discussed internally and before considering to address this with no-loc feature, we will escalate this issue to MSR team if MT engine can address this issue. Will update on this. cc @Rick-Anderson

**Comment by guardrex** [comment]
Thanks @JasonCard ... I'll revert out that one entry in the morning and open an issue to address `blazor.webassembly.js` later.

**Comment by guardrex** [comment]
@Rick-Anderson Although we don't normally code fence file names, how about I code fence that one temporarily? I can revert the code fencing when it's resolved for the build system later.

**Comment by guardrex** [comment]
@JasonCard Do you know what would happen if a zero-width space were used there? Would it be treated as a space so that no-loc for "blazor" would stop the translation?

> blazor\&\#8203;.webassembly.js

**Comment by Rick-Anderson** [comment]
> @Rick-Anderson Although we don't normally code fence file names, how about I code fence that one temporarily? I can revert the code fencing when it's resolved for the build system later.

No problem. Doesn't even have to be temporary.  It's not preferred, but it's OK.

**Comment by guardrex** [comment]
There aren't that many references to it ... only five counting one more that I need to address (*blazor.server.js*). Unless Jason thinks a zero-width space will work, I'll just code fence them for now.

**Comment by JasonCard** [comment]
We would have to try it to find out, but my guess is that even when using zero-width space, ".webassembly.js" may still get localized and/or add space after periods. So I agree using inline code fence would be the best solution for now.

**Comment by danroth27** [comment]
> @danroth27 can let us know if an issue should be opened for engineering to provide a Linux instruction there IF Linux really is supported for debugging. For the time being tho WRT you actually getting it to work, no 🎲🎲 dice on that right now ... your only options there are Windows and macOS.

I believe we're tracking the issue with debugging Blazor WebAssembly on Linux with https://github.com/dotnet/aspnetcore/issues/12970.

File: aspnetcore/blazor/debug.md [hunk 1]
```diff
@@ -44,17 +44,53 @@ Debugging requires either of the following browsers:
...
+       * `Shift+Cmd+D` on macOS
+   4\. The **DevTools** tab opens in the browser. **Reselect the app's tab in the browser window.**
+1. In Visual Studio, select **Debug** > **Attach to Process**.
+1. For the **Connection type**, select **Chrome devtools protocol websocket (no authentication)**.
+1. For the **Connection target**, paste in the HTTP address (not the HTTPS address) of the app.
+1. Select **Refresh** to refresh the entries under **Available processes**.
+1. Select the browser process you want to debug and select **Attach**.
+1. In the **Select Code Type** dialog, select the code type for the specific browser you are attaching to (Edge or Chrome) and then select **OK**.
+
+# [.NET Core CLI](#tab/netcore-cli/)
```
[the remark this answers is not in the document]
Comment (danroth27) [reply aspnetcore/blazor/debug.md]: Note that a bunch of the statements on the top of this article are specific to debugging with the browser dev tools. Like on line 18 it says: "*Early* support exists for debugging Blazor WebAssembly apps running on WebAssembly in Chrome." This language probably needs to be broadened to something like: "Early support exists for debugging Blazor WebAssembly using the browser dev tools in Chromium based browsers (Chrome/Edge) and work is in progress to enable debugging with Visual Studio and Visual Studio Code."

File: aspnetcore/blazor/debug.md [hunk 2]
```diff
@@ -44,17 +44,53 @@ Debugging requires either of the following browsers:
 
 ## Procedure
 
-1. Run a Blazor WebAssembly app in `Debug` configuration. Pass the `--configuration Debug` option to the [dotnet run](/dotnet/core/tools/dotnet-run) command: `dotnet run --configuration Debug`.
-1. Access the app in the browser.
-1. Place the keyboard focus on the app, not the developer tools panel. The developer tools panel can be closed when debugging is initiated.
+# [Visual Studio](#tab/visual-studio)
```
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/blazor/debug.md]: >  it needs to have some text that sets expectations that this is not the actual intended user experience

Will do.

> this should come after the in-browser debugging steps

Confused ... does "this" mean the entire VS tab contents? These are two tabs, so do you want the CLI tab first in the tab control?

File: aspnetcore/blazor/debug.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/blazor/debug.md]: Remove the tabs then and stack the coverage with the "Browser dev tools" section (formerly the ".NET Core CLI" tab) first?

[the remark this answers is not in the document]
Comment (danroth27) [reply aspnetcore/blazor/debug.md]: Oh, these are tabs 😝. Let me look at the review page real quick...

File: aspnetcore/blazor/debug.md [hunk 3]
```diff
@@ -44,17 +44,53 @@ Debugging requires either of the following browsers:
...
    * `Shift+Alt+D` on Windows/Linux
    * `Shift+Cmd+D` on macOS
-1. Follow the steps listed on the screen to restart the browser with remote debugging enabled.
-1. Select the following Blazor-specific keyboard shortcut once again to start the debug session:
+
+   If you receive the **Unable to find debuggable browser tab**, see [Enable remote debugging](#enable-remote-debugging).
+   
+   After enabling remote debugging:
+   
+   1\. A new browser window opens. Close the prior window.
```
Comment (danroth27) [inline aspnetcore/blazor/debug.md]: It doesn't look like these are rendering as a nested numbered list:

image

Comment (guardrex) [reply aspnetcore/blazor/debug.md]: 🐀🍔 *Roasted Rat Burgers!*:tm: 🍔🐀

That should've worked. Let me add a line after the last one there on the next commit, and you can let me know if it worked.

Comment (guardrex) [reply aspnetcore/blazor/debug.md]: btw wrt ...

> work is in progress to enable debugging with Visual Studio

These steps worked here. It seems "enabled" ... just not perhaps "fully" enabled.

Comment (guardrex) [reply aspnetcore/blazor/debug.md]: > *Early* support exists for debugging Blazor WebAssembly using the browser dev tools in Chromium-based browsers (Chrome/Edge), and work is in progress to fully enable debugging with Visual Studio and enable debugging with Visual Studio Code.

File: aspnetcore/blazor/debug.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (danroth27) [reply aspnetcore/blazor/debug.md]: OK, I think the tabs are fine, but we don't want the primary debugging experience when using VS to be the attach to process flow. It's just really clunky right now. If we want to add a section at the end of the doc about how you can manually attach VS to the app, that's fine, but it should be at the end of the doc and it should have some text up front that F5 debugging is coming.

File: aspnetcore/blazor/debug.md (hunk 3 above)
Comment (guardrex) [reply aspnetcore/blazor/debug.md]: Line too long ... how about ...

> *Early* support exists for debugging Blazor WebAssembly using the browser dev tools in Chromium-based browsers (Chrome/Edge). Work is in progress to:
>
> * Fully enable debugging in Visual Studio.
> * Enable debugging with Visual Studio Code.

Comment (danroth27) [reply aspnetcore/blazor/debug.md]: That sounds fine. The point here is that most people will expect to just F5 in VS. That's the experience we plan to ship. The process of manually attaching VS is just something you can try out for now until the work on F5 debugging is done.

**Comment by guardrex** [comment]
Ok ... try it now. I moved the Attach to Process bit into a section near the bottom and cross-link it.

**Comment by guardrex** [comment]
I can't see the review site ...... did those added lines fix the lists?

**Comment by danroth27** [comment]
> I can't see the review site ...... did those added lines fix the lists?

Oh, no, there's still rendering as a single paragraph. Sorry, I forgot to check for that! 🤦‍♂

**Comment by guardrex** [comment]
No worries ... I have another way to fix that. I'll take care of that on a new PR. It hasn't gone live yet. I'll ping u on that one to check there.

**Comment by Sm1le291** [comment]
@guardrex , one more PR or it will be available in next release?

**Comment by guardrex** [comment]
@Sm1le291 Will what be available?

**Comment by Sm1le291** [comment]
@guardrex I mean debug on Linux or it doesn't relate to this issue?

**Comment by guardrex** [comment]
I'm not sure because what we found here for `GetLaunchEdgeInstructions` doesn't seem to match what's discussed on https://github.com/dotnet/aspnetcore/issues/12970. I added a note there ... they might say that it *IS* or *IS NOT* related when they see that ...

https://github.com/dotnet/aspnetcore/issues/12970#issuecomment-582650195

**Comment by guardrex** [comment]
If your behavior matches what we're discussing here with `GetLaunchEdgeInstructions`, then I don't see how it can work on Linux at all until that's fixed. However, it's best for engineering to look and comment. I'm just a silly green dinosaur that writes docs here. :smile:

**Comment by Sm1le291** [comment]
@guardrex sorry, I am out of context, I will show it tommorow) But anyways, Edge in method which shows instructions in  Linux sounds strange:)

**Comment by guardrex** [comment]
It's because that's called no matter what platform you're on when the `Unable to find debuggable browser tab` webpage is thrown up ...

https://github.com/dotnet/aspnetcore/blob/e72223eaf58a3ee6660a922064d2449e47b78253/src/Components/Blazor/Server/src/MonoDebugProxy/BlazorMonoDebugProxyAppBuilderExtensions.cs#L197

**Comment by thetownfool** [comment]
this sounds correct @guardrex the Unknown OS issue appears to be masking an underlying issue. I'm about to patch BlazorMonoDebugProxyAppBuilderExtensions.cs to return a meaningful string, instead of throwing an exception so I can get past this exception and see what i can do about the debuggable port.

**Pull request merged**